{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "# scATAC-seq preprcoes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import snapatac2 as snap\n",
    "snap.__version__\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Fragments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exmpale fragment file, generated by cellranger\n",
    "fragment_file = \"/path/to/fragments.tsv.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for arabidopsis\n",
    "\n",
    "chrom_size = {\n",
    "\"Chr1\":    30427671,\n",
    "\"Chr2\":    19698289,\n",
    "\"Chr3\":    23459830,\n",
    "\"Chr4\":    18585056,\n",
    "\"Chr5\":    26975502,\n",
    "\"ChrM\":    366924,\n",
    "\"ChrC\":    154478\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import data, and all the changed data are recorded in the file\n",
    "data = snap.pp.import_data(\n",
    "    fragment_file,\n",
    "    chrom_sizes=chrom_size,\n",
    "    file=\"scatac.h5ad\",  # Optional\n",
    "    sorted_by_barcode=False,\n",
    "    chrM=[\"ChrC\", \"ChrM\"]\n",
    ")\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Quality Control\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap.pl.frag_size_distr(data, interactive=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GTF file for arabidopsis\n",
    "gtf_file = \"/path/to/at_genes.gtf.gz\"\n",
    "\n",
    "# Create snap object\n",
    "snap.metrics.tsse(data, gtf_file, exclude_chroms=[\"ChrC\",\"ChrM\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap.pl.tsse(data, interactive=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter cell by tsse plot\n",
    "snap.pp.filter_cells(data, min_counts=1500, min_tsse=3.2, max_counts=50000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fragment information\n",
    "data.obsm['fragment_paired'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# close the data\n",
    "data.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tile X Cell matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = snap.read(\"scatac.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap.pp.add_tile_matrix(data, bin_size=100, exclude_chroms=[\"ChrC\",\"ChrM\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A new column named \"selected\" has been created in var.\n",
    "snap.pp.select_features(data, n_features=250000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.var['selected'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap.tl.spectral(data)\n",
    "snap.tl.umap(data)\n",
    "snap.pp.knn(data)\n",
    "snap.tl.leiden(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap.pl.umap(data, color='leiden', interactive=False, height=500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# close the data\n",
    "data.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene activity X Cell Matrix\n",
    "\n",
    "Import the cell-by-bin matrix of single-cell ATAC-seq data and use the snap.pp.make_gene_matrix to generate gene activity matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = snap.read(\"scatac\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = './at_genes.gtf.gz'\n",
    "gene_matrix = snap.pp.make_gene_matrix(data, gtf_file, upstream=2000,gene_name_key=\"gene_id\")\n",
    "gene_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_matrix.write(\"gene_mat.h5ad\", compression='gzip')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Peak X Cell Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = snap.read(\"scatac.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap.tl.macs3(data, groupby='leiden')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = snap.tl.merge_peaks(data.uns['macs3'], chrom_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build peak x cell matrix\n",
    "peak_mat = snap.pp.make_peak_matrix(data, use_rep=peaks['Peaks'])\n",
    "peak_mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write matrix\n",
    "peak_mat.write(\"scatac_peak_mat.h5ad\", compression='gzip')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# close the data\n",
    "data.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## scRNA-seq preprocess\n",
    "\n",
    "This is merely a basic analysis workflow, not necessarily the optimal one. The key lies in annotating the cell types â€” specifically, having a column named cell_type, which is essential for assisting in the subsequent annotation of scATAC-seq data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Set global figure and verbosity parameters\n",
    "sc.set_figure_params(figsize=(6, 6), dpi=100)\n",
    "sc.settings.verbosity = 3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Usually from 10x Genomics format (.h5 or matrix directory)\n",
    "adata = sc.read_10x_h5(\"filtered_feature_bc_matrix.h5\")  # or sc.read_10x_mtx()\n",
    "adata.var_names_make_unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add QC metrics\n",
    "adata.obs['n_counts'] = adata.X.sum(axis=1).A1  # Total transcript count per cell\n",
    "adata.obs['n_genes'] = (adata.X > 0).sum(axis=1).A1  # Number of genes expressed per cell\n",
    "adata.obs['pct_mito'] = (\n",
    "    adata[:, adata.var_names.str.startswith('ATMG')].X.sum(axis=1).A1 / adata.obs['n_counts']\n",
    ") * 100  # Mitochondrial gene percentage\n",
    "\n",
    "# Visualize QC metrics\n",
    "sc.pl.violin(adata, ['n_counts', 'n_genes', 'pct_mito'], jitter=0.4, multi_panel=True)\n",
    "\n",
    "# Filter cells and genes\n",
    "adata = adata[adata.obs['n_genes'] > 200, :]\n",
    "adata = adata[adata.obs['pct_mito'] < 10, :]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers[\"counts\"] = adata.X.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalizing to median total counts\n",
    "sc.pp.normalize_total(adata)\n",
    "# Logarithmize the data\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata, n_top_genes=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highly_variable_genes(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata)\n",
    "sc.pp.neighbors(adata)\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using the igraph implementation and a fixed number of iterations can be significantly faster, especially for larger datasets\n",
    "sc.tl.leiden(adata, flavor=\"igraph\", n_iterations=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'n_counts', 'pct_mito'])  # Visualize clusters and QC metrics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon')  # Differential expression per cluster\n",
    "sc.pl.rank_genes_groups(adata, n_genes=20, sharey=False)  # Plot top marker genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# celltype annotation based on marker genes\n",
    "# this step should be done by human experts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"scrna.h5ad\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "single-cell",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
