{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data1/xhc_data/micromamba/envs/SAMap/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from samap.mapping import SAMAP\n",
    "from samap.analysis import (get_mapping_scores, GenePairFinder,\n",
    "                            sankey_plot, chord_plot, CellTypeTriangles, \n",
    "                            ParalogSubstitutions, FunctionalEnrichment,\n",
    "                            convert_eggnog_to_homologs, GeneTriangles)\n",
    "from samalg import SAM\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy.external as sce\n",
    "from samalg import SAM #import SAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "Not updating the manifold...\n",
      "24238 `os` gene symbols match between the datasets and the BLAST graph.\n",
      "24826 `at` gene symbols match between the datasets and the BLAST graph.\n",
      "59992 `pt` gene symbols match between the datasets and the BLAST graph.\n",
      "24233 `ne` gene symbols match between the datasets and the BLAST graph.\n",
      "24187 `af` gene symbols match between the datasets and the BLAST graph.\n",
      "55519 `sm` gene symbols match between the datasets and the BLAST graph.\n",
      "22838 `bmg` gene symbols match between the datasets and the BLAST graph.\n",
      "22680 `cw` gene symbols match between the datasets and the BLAST graph.\n",
      "32957 `mt` gene symbols match between the datasets and the BLAST graph.\n",
      "32890 `mzei` gene symbols match between the datasets and the BLAST graph.\n",
      "23913 `sl` gene symbols match between the datasets and the BLAST graph.\n",
      "98415 `ta` gene symbols match between the datasets and the BLAST graph.\n",
      "28997 `wm` gene symbols match between the datasets and the BLAST graph.\n",
      "63493 `yc` gene symbols match between the datasets and the BLAST graph.\n",
      "27001 `zm` gene symbols match between the datasets and the BLAST graph.\n"
     ]
    }
   ],
   "source": [
    "# 加载已经处理的SAM\n",
    "sam1 = SAM()\n",
    "sam1.load(\"./result/os.pkl\")\n",
    "\n",
    "sam2 = SAM()\n",
    "sam2.load(\"./result/at_pct20.pkl\")\n",
    "\n",
    "sam3 = SAM()\n",
    "sam3.load(\"./result/pt.pkl\")\n",
    "\n",
    "sam4 = SAM()\n",
    "sam4.load(\"./result/ne.pkl\")\n",
    "\n",
    "sam5 = SAM()\n",
    "sam5.load(\"./result/af.pkl\")\n",
    "\n",
    "sam6 = SAM()\n",
    "sam6.load(\"./result/sm.pkl\")\n",
    "\n",
    "\n",
    "sam7 = SAM()\n",
    "sam7.load(\"./result/bmg.pkl\")\n",
    "\n",
    "sam8 = SAM()\n",
    "sam8.load(\"./result/cw.pkl\")\n",
    "\n",
    "sam9 = SAM()\n",
    "sam9.load(\"./result/mt.pkl\")\n",
    "\n",
    "sam10 = SAM()\n",
    "sam10.load(\"./result/mzei.pkl\")\n",
    "\n",
    "sam11 = SAM()\n",
    "sam11.load(\"./result/sl.pkl\")\n",
    "\n",
    "sam12 = SAM()\n",
    "sam12.load(\"./result/ta.pkl\")\n",
    "\n",
    "sam13 = SAM()\n",
    "sam13.load(\"./result/wm.pkl\")\n",
    "\n",
    "sam14 = SAM()\n",
    "sam14.load(\"./result/yc.pkl\")\n",
    "\n",
    "sam15 = SAM()\n",
    "sam15.load(\"./result/zm.pkl\")\n",
    "\n",
    "# 第一次运行\n",
    "sams = {'os':sam1,'at':sam2,'pt':sam3,'ne':sam4,'af':sam5,\n",
    "        'sm':sam6,'bmg':sam7,'cw':sam8,'mt':sam9,'mzei':sam10,\n",
    "        'sl':sam11,'ta':sam12,'wm':sam13,'yc':sam14,'zm':sam15}\n",
    "\n",
    "sm = SAMAP(\n",
    "        sams,\n",
    "        f_maps = 'data/pep/maps/',\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open(\"result/sm_zhongjian20.pkl\", \"wb\") as f:\n",
    "    pickle.dump(sm,f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open('result/sm_zhongjian20.pkl','rb') as f:\n",
    "    sm = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prepping datasets for translation.\n",
      "Translating feature spaces pairwise.\n",
      "Projecting data into joint latent space. 6424.1041967868805\n",
      "Correcting data with means. 3755.7156360149384\n",
      "Expanding neighbourhoods of species os...\n",
      "Expanding neighbourhoods of species at...\n",
      "Expanding neighbourhoods of species pt...\n",
      "Expanding neighbourhoods of species ne...\n",
      "Expanding neighbourhoods of species af...\n",
      "Expanding neighbourhoods of species sm...\n",
      "Expanding neighbourhoods of species bmg...\n",
      "Expanding neighbourhoods of species cw...\n",
      "Expanding neighbourhoods of species mt...\n",
      "Expanding neighbourhoods of species mzei...\n",
      "Expanding neighbourhoods of species sl...\n",
      "Expanding neighbourhoods of species ta...\n",
      "Expanding neighbourhoods of species wm...\n",
      "Expanding neighbourhoods of species yc...\n",
      "Expanding neighbourhoods of species zm...\n",
      "Indegree coarsening\n",
      "0/10 (0, 184481)\n",
      "1/10 (20000, 184481)\n",
      "2/10 (40000, 184481)\n",
      "3/10 (60000, 184481)\n",
      "4/10 (80000, 184481)\n",
      "5/10 (100000, 184481)\n",
      "6/10 (120000, 184481)\n",
      "7/10 (140000, 184481)\n",
      "8/10 (160000, 184481)\n",
      "9/10 (180000, 184481)\n",
      "Rescaling edge weights by expression correlations.\n",
      "Concatenating SAM objects...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data1/xhc_data/micromamba/envs/SAMap/lib/python3.9/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration 1 complete.\n",
      "Alignment scores:\n",
      "            af        at       bmg        cw        mt      mzei        ne  \\\n",
      "af    0.000000  0.586843  0.625265  0.530045  0.573475  0.610907  0.602264   \n",
      "at    0.529543  0.000000  0.715763  0.698097  0.644411  0.498281  0.491070   \n",
      "bmg   0.508204  0.695273  0.000000  0.582712  0.697811  0.509267  0.520518   \n",
      "cw    0.509776  0.720487  0.639186  0.000000  0.561345  0.491339  0.501987   \n",
      "mt    0.528888  0.718830  0.760025  0.582957  0.000000  0.562971  0.540676   \n",
      "mzei  0.574999  0.523956  0.584525  0.493211  0.529027  0.000000  0.607775   \n",
      "ne    0.531116  0.498271  0.557073  0.492666  0.523408  0.620360  0.000000   \n",
      "os    0.537412  0.679538  0.701437  0.591853  0.606402  0.539762  0.547411   \n",
      "pt    0.547281  0.555711  0.608133  0.477708  0.569510  0.575715  0.539995   \n",
      "sl    0.607279  0.744428  0.731191  0.664372  0.653076  0.557254  0.534864   \n",
      "sm    0.598310  0.598599  0.618160  0.528007  0.567699  0.563152  0.558085   \n",
      "ta    0.542021  0.595253  0.583768  0.528491  0.545942  0.548503  0.505623   \n",
      "wm    0.482481  0.524260  0.584476  0.528757  0.510155  0.471125  0.485512   \n",
      "yc    0.545473  0.605426  0.660261  0.603517  0.604203  0.595199  0.599504   \n",
      "zm    0.528445  0.589869  0.617884  0.528068  0.548459  0.514093  0.471543   \n",
      "\n",
      "            os        pt        sl        sm        ta        wm        yc  \\\n",
      "af    0.575816  0.538591  0.576348  0.593742  0.499033  0.498260  0.589005   \n",
      "at    0.621715  0.438517  0.641238  0.496019  0.477274  0.442018  0.586294   \n",
      "bmg   0.616092  0.487022  0.605219  0.529122  0.446511  0.494492  0.610332   \n",
      "cw    0.545242  0.456582  0.640249  0.495323  0.474838  0.514107  0.594345   \n",
      "mt    0.629045  0.501362  0.607906  0.556746  0.501146  0.494197  0.615780   \n",
      "mzei  0.571609  0.503784  0.562469  0.518827  0.488026  0.456004  0.553602   \n",
      "ne    0.476862  0.468514  0.533975  0.524889  0.445012  0.480587  0.584843   \n",
      "os    0.000000  0.484844  0.592207  0.546257  0.590775  0.483927  0.618156   \n",
      "pt    0.538577  0.000000  0.552875  0.545118  0.499501  0.477638  0.540248   \n",
      "sl    0.657457  0.530378  0.000000  0.587357  0.503415  0.565481  0.624540   \n",
      "sm    0.584038  0.528618  0.602319  0.000000  0.487809  0.514379  0.583160   \n",
      "ta    0.602913  0.499996  0.594224  0.550952  0.000000  0.459817  0.537867   \n",
      "wm    0.511281  0.475038  0.533807  0.492907  0.447045  0.000000  0.546091   \n",
      "yc    0.567838  0.531933  0.603105  0.562787  0.501531  0.532248  0.000000   \n",
      "zm    0.603313  0.479684  0.567211  0.562259  0.517128  0.521654  0.545689   \n",
      "\n",
      "            zm  \n",
      "af    0.540287  \n",
      "at    0.534131  \n",
      "bmg   0.486180  \n",
      "cw    0.530075  \n",
      "mt    0.547364  \n",
      "mzei  0.487977  \n",
      "ne    0.471220  \n",
      "os    0.552877  \n",
      "pt    0.478745  \n",
      "sl    0.613666  \n",
      "sm    0.556021  \n",
      "ta    0.583156  \n",
      "wm    0.538117  \n",
      "yc    0.539291  \n",
      "zm    0.000000  \n",
      "Calculating gene-gene correlations in the homology graph...\n",
      "Prepping datasets for translation.\n",
      "Translating feature spaces pairwise.\n",
      "Projecting data into joint latent space. 2816.6419506073\n",
      "Correcting data with means. 2679.8758642673492\n",
      "Expanding neighbourhoods of species os...\n",
      "Expanding neighbourhoods of species at...\n",
      "Expanding neighbourhoods of species pt...\n",
      "Expanding neighbourhoods of species ne...\n",
      "Expanding neighbourhoods of species af...\n",
      "Expanding neighbourhoods of species sm...\n",
      "Expanding neighbourhoods of species bmg...\n",
      "Expanding neighbourhoods of species cw...\n",
      "Expanding neighbourhoods of species mt...\n",
      "Expanding neighbourhoods of species mzei...\n",
      "Expanding neighbourhoods of species sl...\n",
      "Expanding neighbourhoods of species ta...\n",
      "Expanding neighbourhoods of species wm...\n",
      "Expanding neighbourhoods of species yc...\n",
      "Expanding neighbourhoods of species zm...\n",
      "Indegree coarsening\n",
      "0/10 (0, 184481)\n",
      "1/10 (20000, 184481)\n",
      "2/10 (40000, 184481)\n",
      "3/10 (60000, 184481)\n",
      "4/10 (80000, 184481)\n",
      "5/10 (100000, 184481)\n",
      "6/10 (120000, 184481)\n",
      "7/10 (140000, 184481)\n",
      "8/10 (160000, 184481)\n",
      "9/10 (180000, 184481)\n",
      "Rescaling edge weights by expression correlations.\n",
      "Concatenating SAM objects...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data1/xhc_data/micromamba/envs/SAMap/lib/python3.9/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration 2 complete.\n",
      "Alignment scores:\n",
      "            af        at       bmg        cw        mt      mzei        ne  \\\n",
      "af    0.000000  0.686574  0.718728  0.630791  0.656302  0.689279  0.683775   \n",
      "at    0.591064  0.000000  0.794599  0.746092  0.722606  0.595787  0.599694   \n",
      "bmg   0.582643  0.782171  0.000000  0.651994  0.757094  0.584719  0.612353   \n",
      "cw    0.562948  0.765785  0.695283  0.000000  0.627848  0.547255  0.566775   \n",
      "mt    0.588614  0.794229  0.810476  0.650506  0.000000  0.636187  0.622583   \n",
      "mzei  0.640339  0.658535  0.699904  0.587126  0.628416  0.000000  0.699602   \n",
      "ne    0.587292  0.600521  0.638538  0.552991  0.595629  0.674535  0.000000   \n",
      "os    0.594508  0.739427  0.768577  0.651604  0.694915  0.612501  0.628743   \n",
      "pt    0.619514  0.666115  0.697197  0.544649  0.659018  0.650775  0.604104   \n",
      "sl    0.657112  0.816037  0.799233  0.717211  0.703471  0.602980  0.602307   \n",
      "sm    0.664583  0.715787  0.709477  0.611046  0.645874  0.635047  0.627605   \n",
      "ta    0.598370  0.704448  0.660043  0.604489  0.622039  0.625906  0.568564   \n",
      "wm    0.563237  0.650818  0.678444  0.609237  0.595270  0.528705  0.551610   \n",
      "yc    0.609769  0.676186  0.719117  0.656425  0.661641  0.642414  0.650501   \n",
      "zm    0.581843  0.666145  0.688258  0.587433  0.603393  0.580794  0.533702   \n",
      "\n",
      "            os        pt        sl        sm        ta        wm        yc  \\\n",
      "af    0.665081  0.623150  0.647578  0.679579  0.591949  0.597801  0.677750   \n",
      "at    0.699610  0.517466  0.723527  0.618231  0.580693  0.526582  0.675865   \n",
      "bmg   0.713889  0.554500  0.679346  0.613441  0.520844  0.566651  0.681159   \n",
      "cw    0.586405  0.515441  0.691597  0.563406  0.535759  0.560870  0.649364   \n",
      "mt    0.703268  0.583634  0.667540  0.620463  0.573912  0.563924  0.682597   \n",
      "mzei  0.669332  0.557896  0.628045  0.629400  0.606692  0.519498  0.632437   \n",
      "ne    0.549290  0.503601  0.567858  0.584301  0.521383  0.529096  0.640262   \n",
      "os    0.000000  0.568626  0.653415  0.633466  0.655974  0.543875  0.679804   \n",
      "pt    0.617705  0.000000  0.606411  0.620837  0.585906  0.545260  0.609482   \n",
      "sl    0.714699  0.581322  0.000000  0.666059  0.576382  0.607607  0.687739   \n",
      "sm    0.672788  0.604867  0.672187  0.000000  0.603847  0.580193  0.647460   \n",
      "ta    0.668683  0.561429  0.669927  0.656856  0.000000  0.521652  0.602141   \n",
      "wm    0.593910  0.530707  0.606645  0.558337  0.533323  0.000000  0.615722   \n",
      "yc    0.620733  0.568465  0.649129  0.607891  0.553308  0.569868  0.000000   \n",
      "zm    0.681789  0.530520  0.619217  0.637226  0.594573  0.591889  0.596870   \n",
      "\n",
      "            zm  \n",
      "af    0.582824  \n",
      "at    0.580415  \n",
      "bmg   0.539180  \n",
      "cw    0.567502  \n",
      "mt    0.594021  \n",
      "mzei  0.564624  \n",
      "ne    0.506364  \n",
      "os    0.587358  \n",
      "pt    0.515748  \n",
      "sl    0.659469  \n",
      "sm    0.625449  \n",
      "ta    0.643751  \n",
      "wm    0.612547  \n",
      "yc    0.560986  \n",
      "zm    0.000000  \n",
      "Calculating gene-gene correlations in the homology graph...\n",
      "Prepping datasets for translation.\n",
      "Translating feature spaces pairwise.\n",
      "Projecting data into joint latent space. 2659.232491493225\n",
      "Correcting data with means. 2778.1549994945526\n",
      "Expanding neighbourhoods of species os...\n",
      "Expanding neighbourhoods of species at...\n",
      "Expanding neighbourhoods of species pt...\n",
      "Expanding neighbourhoods of species ne...\n",
      "Expanding neighbourhoods of species af...\n",
      "Expanding neighbourhoods of species sm...\n",
      "Expanding neighbourhoods of species bmg...\n",
      "Expanding neighbourhoods of species cw...\n",
      "Expanding neighbourhoods of species mt...\n",
      "Expanding neighbourhoods of species mzei...\n",
      "Expanding neighbourhoods of species sl...\n",
      "Expanding neighbourhoods of species ta...\n",
      "Expanding neighbourhoods of species wm...\n",
      "Expanding neighbourhoods of species yc...\n",
      "Expanding neighbourhoods of species zm...\n",
      "Indegree coarsening\n",
      "0/10 (0, 184481)\n",
      "1/10 (20000, 184481)\n",
      "2/10 (40000, 184481)\n",
      "3/10 (60000, 184481)\n",
      "4/10 (80000, 184481)\n",
      "5/10 (100000, 184481)\n",
      "6/10 (120000, 184481)\n",
      "7/10 (140000, 184481)\n",
      "8/10 (160000, 184481)\n",
      "9/10 (180000, 184481)\n",
      "Rescaling edge weights by expression correlations.\n",
      "Concatenating SAM objects...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data1/xhc_data/micromamba/envs/SAMap/lib/python3.9/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration 3 complete.\n",
      "Alignment scores:\n",
      "            af        at       bmg        cw        mt      mzei        ne  \\\n",
      "af    0.000000  0.710468  0.731669  0.654319  0.672322  0.706292  0.699957   \n",
      "at    0.608467  0.000000  0.803459  0.755812  0.731871  0.621541  0.616263   \n",
      "bmg   0.608502  0.790670  0.000000  0.664290  0.764614  0.599508  0.628533   \n",
      "cw    0.574393  0.770474  0.701947  0.000000  0.635139  0.562932  0.579198   \n",
      "mt    0.617033  0.804213  0.814369  0.660116  0.000000  0.648848  0.638250   \n",
      "mzei  0.651362  0.677287  0.714416  0.612552  0.650434  0.000000  0.713487   \n",
      "ne    0.604678  0.618008  0.651709  0.568927  0.610971  0.682036  0.000000   \n",
      "os    0.611541  0.748542  0.774907  0.662591  0.703550  0.626244  0.641735   \n",
      "pt    0.643503  0.693651  0.706886  0.564832  0.675918  0.655397  0.615416   \n",
      "sl    0.677209  0.823168  0.804595  0.724194  0.709110  0.605861  0.617425   \n",
      "sm    0.676869  0.729975  0.723609  0.624222  0.657843  0.646612  0.638194   \n",
      "ta    0.609565  0.713409  0.666984  0.618067  0.630552  0.631435  0.579046   \n",
      "wm    0.585010  0.674516  0.693344  0.624141  0.609467  0.562612  0.572518   \n",
      "yc    0.618460  0.681077  0.722026  0.664577  0.666414  0.646620  0.656547   \n",
      "zm    0.588894  0.682668  0.697559  0.597366  0.611409  0.589998  0.549947   \n",
      "\n",
      "            os        pt        sl        sm        ta        wm        yc  \\\n",
      "af    0.689128  0.641360  0.665093  0.693907  0.606938  0.619025  0.697770   \n",
      "at    0.714991  0.541566  0.733239  0.649172  0.595655  0.557308  0.684949   \n",
      "bmg   0.731275  0.572013  0.691516  0.634610  0.536333  0.584873  0.690394   \n",
      "cw    0.596885  0.535035  0.699184  0.582115  0.549177  0.572642  0.656268   \n",
      "mt    0.712799  0.598888  0.676538  0.636292  0.587913  0.586178  0.690103   \n",
      "mzei  0.681576  0.574870  0.637286  0.649002  0.618133  0.548655  0.643786   \n",
      "ne    0.568536  0.513408  0.579277  0.596753  0.537818  0.540795  0.649372   \n",
      "os    0.000000  0.588993  0.663839  0.651760  0.666345  0.568981  0.687396   \n",
      "pt    0.636899  0.000000  0.620818  0.641497  0.604752  0.557834  0.617203   \n",
      "sl    0.729127  0.595989  0.000000  0.680049  0.585066  0.620366  0.697273   \n",
      "sm    0.689790  0.618556  0.682414  0.000000  0.626091  0.594865  0.659710   \n",
      "ta    0.674939  0.574933  0.676886  0.675440  0.000000  0.540499  0.609724   \n",
      "wm    0.604713  0.549209  0.628009  0.586575  0.556411  0.000000  0.631586   \n",
      "yc    0.630456  0.574345  0.655276  0.623312  0.563211  0.583047  0.000000   \n",
      "zm    0.695007  0.541121  0.633240  0.649338  0.609522  0.599788  0.609935   \n",
      "\n",
      "            zm  \n",
      "af    0.586934  \n",
      "at    0.587352  \n",
      "bmg   0.556750  \n",
      "cw    0.570959  \n",
      "mt    0.603588  \n",
      "mzei  0.582854  \n",
      "ne    0.516347  \n",
      "os    0.593113  \n",
      "pt    0.523451  \n",
      "sl    0.668250  \n",
      "sm    0.642943  \n",
      "ta    0.653596  \n",
      "wm    0.626273  \n",
      "yc    0.561766  \n",
      "zm    0.000000  \n",
      "Running UMAP on the stitched manifolds.\n",
      "Elapsed time: 676.8429614305496 minutes.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<samalg.SAM at 0x7fe3bb47edc0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "# 在其他需要OpenBLAS的库导入前设置\n",
    "os.environ[\"OPENBLAS_NUM_THREADS\"] = \"10\"\n",
    "\n",
    "# run the SAMap algorithm using the run function.\n",
    "sm.run(pairwise=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save results to file\n",
    "import pickle\n",
    "with open('result/15_samap.pkl','wb') as f:\n",
    "    pickle.dump(sm,f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
